Allow an InputSource to declare a run/lumi is the last to be merged

wddgit opened the pull request:
#### PR description:

Allow an InputSource to declare that a run entry or a lumi entry is the last one that needs to be merged. The next run entry should have a different run number or ProcessHistoryID. Or the next lumi entry should be associated with a different run or have a different luminosity block number. Or the next run or lumi entry could be in a different file.

It is OK for InputSources not to make this declaration. The Framework can and already does figure this out by examining the next entry returned by getNextItemType. Existing sources should work OK.

This might be a useful optimization in an online source when the getNextItemType function takes a long time to return. If the last item is declared, then the global run transition or the global lumi transition can begin immediately. Otherwise, the Framework has to wait for getNextItemType to return for the next entry in order to know that the previous entry was the last. And that delays global begin run or global begin lumi.

This probably will not help much for offline input sources because getNextItemType returns quickly.

The Framework will detect the case where a source falsely declares an item is the last run or lumi entry. It will throw an exception indicating a bug in the InputSource.

This delay was noticed when investigating issue #42931, but this PR does not address either of the problems listed there.

There are some minor changes in the InputSource interface. The getNextItemType function now returns an object of type ```ItemTypeInfo``` instead of the enum ItemType. Also, the ItemType enum is changed to be an ```enum class``` (more of a style modernization, I could back that out if requested).

#### PR validation:

Three new unit tests are added for this feature. Existing core unit tests pass. Limited runTheMatrix passes.

Manually, you can also see the delay vanishes in the log file of one of the unit tests, but it is difficult to turn that into a pass/fail criteria as the timing of execution can vary and we don't want tests that occasionally fail due to timing variations.

wddgit committed:
Allow an InputSource to declare a run/lumi entry is the last to be merged

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43457/38008

- This PR adds an extra 152KB to repository

wddgit commented:
please test

cmsbuild commented:
-1

**Failed Tests**: AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-274562/36219/summary.html
**COMMIT**: 7b27eaec55b6f857727ff4ec85c1867d0aeb5abe
**CMSSW**: CMSSW_14_0_X_2023-11-30-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43457/36219/install.sh` to create a dev area with all the needed externals and cmssw changes.


## AddOn Tests

- [hlt_data_GRun](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-274562/36219/addOnTests/hlt_data_GRun)
```
----- Begin Fatal Exception 01-Dec-2023 00:47:00 CET-----------------------
An exception of category 'StdException' occurred while
   [0] Calling EventProcessor::runToCompletion (which does almost everything after beginJob and before endJob)
Exception Message:
A std::exception was thrown.
vector::_M_range_check: __n (which is 0) >= this->size() (which is 0)
   Additional Info:
      [a] Another exception was caught while trying to clean up runs after the primary fatal exception.
----- End Fatal Exception -------------------------------------------------
```
- [hlt_data_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-274562/36219/addOnTests/hlt_data_HIon)
```
----- Begin Fatal Exception 01-Dec-2023 00:47:03 CET-----------------------
An exception of category 'StdException' occurred while
   [0] Calling EventProcessor::runToCompletion (which does almost everything after beginJob and before endJob)
Exception Message:
A std::exception was thrown.
vector::_M_range_check: __n (which is 0) >= this->size() (which is 0)
   Additional Info:
      [a] Another exception was caught while trying to clean up runs after the primary fatal exception.
----- End Fatal Exception -------------------------------------------------
```
- [hlt_data_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-274562/36219/addOnTests/hlt_data_PIon)
```
----- Begin Fatal Exception 01-Dec-2023 00:46:43 CET-----------------------
An exception of category 'StdException' occurred while
   [0] Calling EventProcessor::runToCompletion (which does almost everything after beginJob and before endJob)
Exception Message:
A std::exception was thrown.
vector::_M_range_check: __n (which is 0) >= this->size() (which is 0)
   Additional Info:
      [a] Another exception was caught while trying to clean up runs after the primary fatal exception.
----- End Fatal Exception -------------------------------------------------
```


## Comparison Summary



**Summary**:
 - You potentially [removed 431 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-30-1100+274562/60066/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 13 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-30-1100+274562/60066/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3370032
 - DQMHistoTests: [Total failures: 95](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-30-1100+274562/60066/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3369915
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-30-1100+274562/60066/triggerResults): no differences found

wddgit commented:
please test

Try again. See if the failures are reproducible. The only thing I noticed so far is that the 3 failures occurred within 20 seconds of each other in real time. Maybe some system glitch...

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43457/38041

- This PR adds an extra 152KB to repository

wddgit commented:
please test

Try again. I missed adding ```ServiceRegistry::Operate operate(serviceToken_);``` in a spot where it was needed. I squashed and force pushed a fix.

cmsbuild commented:
Pull request #43457 was updated. @Dr15Jones, @smuzaffar, @makortel, @syuvivida, @cmsbuild, @rvenditti, @tjavaid, @antoniovagnerini, @nothingface0 can you please check and sign again.

makortel commented on FWCore/Framework/test/unit_test_outputs/test_deepCall_unscheduled.log:
```diff
@@ -109,6 +106,9 @@ ModuleCallingContext state = Running
 ++++ finished: begin run: stream = 0 run = 1 time = 1000000
+++++ queuing: EventSetup synchronization run: 1 lumi: 1 event: 0
+++++ pre: EventSetup synchronizing run: 1 lumi: 1 event: 0
+++++ post: EventSetup synchronizing run: 1 lumi: 1 event: 0
```
Why does order of beginRun and EventSetup synchronization change?

wddgit replied:
First note that the EventSetup signals are for the lumi initialization, not the run initialization. I don't think it technically matters whether they come before or after stream begin run.

This is related to the previous question. Those EventSetup signals fire while ```beginLumiAsync``` is being executed (not while tasks it starts are executed, it's at the top level). We've moved ```beginLumiAsync``` inside the ```then``` part of the new chain so it does not execute while ```handleNextItemAfterMergingRunEntries``` executes at the top level anymore, it is now inside a separate task that it creates. There is intentionally synchronization in there to get stream begin run to execute before global begin lumi in the single threaded case. This preserved long standing behavior in the single threaded case. In the multi threaded case, it is possible for stream end run and the EventSetup signals (and global begin lumi) to all execute concurrently.

I like this new ordering in the single threaded case. It puts the lumi EventSetup signals closer to global begin lumi. Although it was just an accidental side effect of the changes.

makortel commented on FWCore/Integration/plugins/SourceWithWaits.cc:
```diff
@@ -63,98 +63,220 @@ namespace edmtest {
-    } else if (currentLumi_ == 0u) {
+      if (currentRun_ != multipleEntriesForRun_) {
+        startedNewRun_ = true;
+        return (declareLast_ || declareAllLast_)
+                   ? edm::InputSource::ItemTypeInfo(edm::InputSource::ItemType::IsRun,
+                                                    edm::InputSource::ItemPosition::LastItemToBeMerged)
+                   : edm::InputSource::ItemTypeInfo(edm::InputSource::ItemType::IsRun,
+                                                    edm::InputSource::ItemPosition::NotLastItemToBeMerged);
```
I suppose this could be also
```suggestion
        return edm::InputSource::ItemTypeInfo(edm::InputSource::ItemType::IsRun,
                                                    (declareLast_ || declareAllLast_) ? edm::InputSource::ItemPosition::LastItemToBeMerged : edm::InputSource::ItemPosition::NotLastItemToBeMerged);
```
or
```suggestion
        auto const position = (declareLast_ || declareAllLast_) ? edm::InputSource::ItemPosition::LastItemToBeMerged : edm::InputSource::ItemPosition::NotLastItemToBeMerged);
        return edm::InputSource::ItemTypeInfo(edm::InputSource::ItemType::IsRun, position);
```

I'm not really sure which one to prefer though.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-274562/36268/summary.html
**COMMIT**: a4df6b8a23c6e59c6967df25648e24984b75eb80
**CMSSW**: CMSSW_14_0_X_2023-12-01-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43457/36268/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 65 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-01-1100+274562/60105/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 24 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-01-1100+274562/60105/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3370032
 - DQMHistoTests: [Total failures: 1200](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-01-1100+274562/60105/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3368810
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-01-1100+274562/60105/triggerResults): no differences found

wddgit commented on FWCore/Framework/src/EventProcessor.cc:
```diff
@@ -2131,25 +2155,53 @@ namespace edm {
-        WaitingTaskHolder copyHolder(iHolder);
-        iHolder.doneWaiting(std::current_exception());
-        endRunAsync(std::move(iRunStatus), std::move(iHolder));
+    chain::first([this, iRunStatus](auto nextTask) {
```
The main action in the task in the ```first``` part of the chain is to call ```nextTransitionType```. Before the changes in this PR, this call always occurred in the function ```readAndMergeRunEntriesAsync```, which is contained in the ```first``` part of a different chain which contains the call to ```handleNextItemAfterMergingRunEntries``` in one of its ```then``` parts. It was always sequenced to call ```nextTransitionType``` before we reached this part of the code.

After the changes in this PR, there are two cases.

* In the first case the input source does not use the new feature declaring a run or lumi entry as the last to be merged. In this case, ```needToCallNext_``` is always false. The code in the ```first``` section does nothing. There is no benefit to having the chain in this case.
* In the other case, the Framework needs to know what is next before it knows whether to start a new lumi, end the run, start a new run, or pause for a file transition. It cannot proceed until ```nextTransitionType``` returns, because it doesn't know what to do yet. In addition, ```nextTransitionType``` interacts with the input source and may execute concurrently with other things interacting with the input source (for example, dereferencing a ```Ptr```). Therefore, we run it as a task in a serial queue (and protect it inside a mutex also). Because it is not running inside the same task and possibly on a different thread, we need the chain to time order completion of ```nextTransitionType``` and the transitions that follow (there are alternative ways to order this, but chain seemed the simplest alternative). That is the main benefit of the chain.

The chain also helps keep the sequence the same as before so that stream begin run executes before global begin lumi in the single threaded case, not really necessary but nice.

makortel replied:
Ah right, I see the need now. The lambda argument `nextTask` is being passed to `beginLumiAsync()` call.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43457/38070

- This PR adds an extra 152KB to repository

cmsbuild commented:
Pull request #43457 was updated. @smuzaffar, @makortel, @tjavaid, @antoniovagnerini, @nothingface0, @cmsbuild, @Dr15Jones, @syuvivida, @rvenditti can you please check and sign again.

wddgit commented:
please test

I force pushed changes base on Matti's last comment. Also removed some harmless but unnecessary ```edm::InputSource``` scope resolution qualifiers. All  in the same file (SourceWithWaits.cc).

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-274562/36325/summary.html
**COMMIT**: 3dcaf1c60dfdc4ea24e45be75aa282895f33ded0
**CMSSW**: CMSSW_14_0_X_2023-12-04-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43457/36325/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 23 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-04-1100+274562/60161/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 6 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-04-1100+274562/60161/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3370032
 - DQMHistoTests: [Total failures: 92](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-04-1100+274562/60161/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3369918
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-04-1100+274562/60161/triggerResults): no differences found

Dr15Jones commented on FWCore/Framework/interface/InputSource.h:
```diff
@@ -349,7 +378,7 @@ namespace edm {
       resetLuminosityBlockAuxiliary();
       resetRunAuxiliary();
-      state_ = IsInvalid;
+      state_ = ItemType::IsInvalid;
```
How about 
```suggestion
      state_ = ItemTypeInfo();
```
?

Dr15Jones commented on FWCore/Framework/src/EventProcessor.cc:
```diff
@@ -2131,25 +2155,53 @@ namespace edm {
-      // if appropriate.
-      endRunAsync(std::move(iRunStatus), std::move(iHolder));
-    }
+    }) | chain::then([this, iRunStatus](std::exception_ptr const* iException, auto nextTask) {
```
If you use `WaitingTaskHolder nextTask` then you have a copy and can call `doneWaiting` explicitly. Then in the `catch` clause you could do `nextTask.presetTaskAsFailed(std::current_exception())`. This would allow you to 
- add a `return` at line 2188,
-  replace the `else if` with `if`
-  remove the call to `endRunAsync` in the `catch` clause
- do away with the `else`

wddgit replied:
Done. Thanks.

Dr15Jones commented on FWCore/Framework/interface/EventProcessor.h:
```diff
@@ -369,7 +369,7 @@ namespace edm {
     bool printDependencies_ = false;
     bool deleteNonConsumedUnscheduledModules_ = true;
-    bool firstItemAfterLumiMerge_ = true;
+    bool needToCallNext_ = true;
```
I'm really not a fan of all the `naked` calls to look at or change the value of `needToCallNext_` in the `EventProcessor.cc` file. Could such behavior be moved to new member functions?

wddgit replied:
Done. Thanks.

wddgit commented on FWCore/Framework/interface/InputSource.h:
```diff
@@ -51,7 +51,36 @@ namespace edm {
+
+    class ItemTypeInfo {
+    public:
+      ItemTypeInfo(ItemType type = ItemType::IsInvalid, ItemPosition position = ItemPosition::Invalid)
```
```constexpr``` works. I'll add that.

```explicit``` does not work unless I explicitly construct an ```ItemInfoType``` in a bunch of places. Here are two patterns that occur repeatedly.

First:
```
return InputSource::ItemType::IsFile;
```
where this is used in a function (e.g. getNextItemType in a source) that returns an ItemTypeInfo and is expecting an implicit conversion from ItemType (the enumeration) to ItemTypeInfo.


Second:
```
state_ = ItemType::IsEvent;
```
Requires an implicit conversion from ItemType to ItemTypeInfo.

I could make the changes required to make explicit work, but the implicit conversion is at least making the code less verbose. I'd prefer not to, but I don't feel strongly about it. Let me know if you really want me to make those changes.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43457/38131

- This PR adds an extra 152KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/interface/EventProcessor.h  modified in PR(s): #43522
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #43522

wddgit commented on FWCore/Framework/src/EventProcessor.cc:
```diff
@@ -2131,25 +2155,53 @@ namespace edm {
+    chain::first([this, iRunStatus](auto nextTask) {
+      if (needToCallNext_) {
+        auto group = nextTask.group();
+        sourceResourcesAcquirer_.serialQueueChain().push(
```
Done. Thanks.

wddgit commented:
I implemented all Chris's suggestion except for making the ItemTypeInfo constructor explicit (see comment above). Thanks, it is better. Also rebased. I'll rerun the tests when code checks passes.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43457/38140

- This PR adds an extra 56KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/interface/EventProcessor.h  modified in PR(s): #43522
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #43522

wddgit commented:
please test

makortel commented:
The usual set of
* https://github.com/cms-sw/cmssw/issues/39803
* https://github.com/cms-sw/cmssw/issues/43456
* https://github.com/cms-sw/cmssw/issues/43293

makortel commented:
+core

makortel commented:
@cms-sw/dqm-l2 Could you review and sign? The changes in your area should be straightforward.

tjavaid commented:
+1

- with spurious differences coming from WFs mentioned above by @makortel .

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @rappoccio, @antoniovilela, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)

antoniovilela commented:
+1
